### PR TITLE
Loading Modal Window

### LINKED ISSUE #328: Frontend: Display loading animation, when TableComponent is loading the next page
If connection is slow, or server is slow, then it's not obvious, that the page has been changed.

### PR DESCRIPTION
**Pull request by sanyavertolet**
This PR closes #328

### What's done:
- [x] Added logo-anim.gif
- [x] Implemented loading handler that is a wrapper for any request
- [x] Reworked FC for role changing
- [x] Reworked FC for tables
- [x] Fixed SAVE format button

### What's left out of scope:
- Fix loading handler for ExecutionView.kt

### CODE REVIEW

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt [hunk 1]
```diff
@@ -127,8 +125,9 @@ class App : ComponentWithScope<PropsWithChildren, AppState>() {
 
     @Suppress("EMPTY_BLOCK_STRUCTURE_ERROR", "TOO_LONG_FUNCTION", "LongMethod")
     override fun RBuilder.render() {
+        // while (true) { }
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt:128]: what does this comment mean?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt [hunk 2]
```diff
@@ -127,8 +125,9 @@ class App : ComponentWithScope<PropsWithChildren, AppState>() {
...
         HashRouter {
-            errorModalHandler {
+            requestModalHandler {
```
Comment (akuleshov7) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt:130]: why changed?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt:128]: Oops, will remove it

Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/App.kt:128]: Used to use it to test changes connected with `index.html` changes

**Comment by akuleshov7** [comment]
fix build and test and lgtm

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/RequestUtils.kt [hunk 3]
```diff
@@ -33,13 +33,18 @@ import kotlinx.serialization.json.Json
...
+interface WithRequestStatusContext {
     /**
      * @param response
      */
     fun setResponse(response: Response)
+
+    /**
+     * @param lambda
+     */
+    fun setLoadingCounter(lambda: (Int) -> Int)
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/RequestUtils.kt:47]:
```suggestion
    fun setLoadingCounter(transform: (oldValue: Int) -> Int)
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/RequestUtils.kt [hunk 4]
```diff
@@ -143,25 +170,56 @@ suspend fun WithRequestStatusContext.delete(
...
+}
+
+/**
+ * Handler that allows to show loading modal
+ *
+ * @param request REST API method
+ * @return [Response] received with [request]
+ */
+@Suppress("MAGIC_NUMBER")
+internal suspend fun Component<*, *>.classLoadingHandler(request: suspend () -> Response): Response {
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/utils/RequestUtils.kt:202]: Why do you have two functions for this? And if this one is intended as a wrapper to be used in components, then why is `Component<*, *>.loadingHandler` not private?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/HasErrorModal.kt [hunk 5]
```diff
@@ -109,3 +139,10 @@ data class ErrorModalState(
     val errorLabel: String,
     val confirmationText: String = "Close",
 )
+
+/**
+ * @property isLoadingModalOpen
+ */
+data class LoadingModalState(
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/HasErrorModal.kt:146]: I wanted all handlers to have the same structure. But it seems that there is no need in this data class. If you think it is better to remove it, I will fix it

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt [hunk 6]
```diff
@@ -863,7 +834,8 @@ class ProjectView : AbstractView<ProjectExecutionRouteProps, ProjectViewState>(f
         val response = get(
             "$apiUrl/latestExecution?name=${state.project.name}&organizationName=${state.project.organization.name}",
             headers,
-            responseHandler = ::noopResponseHandler
+            ::noopLoadingHandler,
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/ProjectView.kt]: Done

**Pull request merged**